### PR TITLE
Use shared_ptr for CNode inside CConnman

### PR DESCRIPTION
**Pull request by willcl-ark**
Switch to using smart pointers to `CNode`s inside of `CConnman`.

Currently we are manually refcounting CNodes which is potentially error-prone and makes operations such as deleting them from multiple threads difficult without introducing new locks or other synchronisation operations (see https://github.com/bitcoin/bitcoin/pull/27912).

Switch to using `std::shared_ptr` references to `CNode`s inside of `m_nodes` and `m_nodes_disconnected` to give us better memory safety today, and in the future allow `AttemptToEvictConnection` (and optionally other sites) to safely synchronously disconnect nodes when needed.

Opening as draft for now as I want to both gauge feedback on the approach, and see which PRs this may conflict with (#27213?) before moving it forwards.

CC @vasild

### CODE REVIEW

**Comment by dergoegge** [comment]
Also see #10738

**Comment by willcl-ark** [comment]
> Also see #10738

Thanks, I hadn't seen this one before. Will take a look tonight

**Comment by vasild** [comment]
> Also see #10738

:+1: 

I guess it must be possible to do that without "reinventing the wheel aka introducing custom smart pointer types", like mentioned in the last comment.

**Review by dergoegge — COMMENTED** [review]
Concept ACK on getting rid of our custom ref counting.

I actually implemented something similar a while ago (https://github.com/dergoegge/bitcoin/commit/b3913b4c052b37790b3737e5ddb9c2d029502477) but didn't pursue because it doesn't retain the behavior we currently have w.r.t. which thread actually deletes the `CNode`. That is also the reason why https://github.com/bitcoin/bitcoin/pull/10738 is a bit more complex and introduces new pointer types.

> which thread actually deletes the CNode.

The reason why that is important is because we only want to call `FinalizeNode` once and we want there to be no other references to that `CNode` at that time.

The changes in this PR and the current ref-counting don't guarantee that, they just aim to work that way but it always requires manual review to avoid bugs.

I think I'd prefer an approach that guarantees (at least conceptually) correct behavior (like e.g. the pointer types in #10738).

**Comment by dergoegge** [comment]
cc @theuni (just in case you are interested in chiming in)

**Comment by willcl-ark** [comment]
Thanks for the conceptual review @dergoegge.

I've forced push a relatively different approach here, trying to take into consideration your review points, and #10738, but trying not to require introducing the decaying pointers. This has resulted in the following conceptual changes:

1. Taking the idea of @theuni's `GetnodesCopy()` in `m_nodes` iterations by re-purposing and upgrading `NodesSnapshot()` a bit, allowing removal of many instances of `m_nodes_mutex` in the code, as he achieved.
2. Made `CNode` a friend of `CConnman`. This allows us to call back to `DeleteNode()` automagically 🪄  on CNode destruction when the final reference is dropped. Then we can just worry about dropping all references, which seems easier to reason about for me. I think it might be possible to modify `DeleteNode()` further to assert no `shared_ptr`s remain for it --ensuring that nobody can accidently re-introduce unsafe deletion in the codebase in the future --, but I haven't looked at this yet.

> In a way this is worse than the custom ref counting because it is harder to assert that we actually don't take new references after this check.

3. I re-worked this part of `DisconnectNodes()` in general, but IMO this is safe-enough to drop from `m_nodes_disconnected` outright, as the only users of `m_nodes_disconnected` are `StopNodes()` and `DisconnectNodes()`. To belt-and-suspenders this anyway, I added a final optional commit which adds an `m_nodes_disconnected_mutex` to ensure no new references can be created while we are dropping from `m_nodes_disconnected`.

The commits need tidying up, rewording etc. still, but otherwise would be curious to know if this new approach looked better to you?

**Comment by theuni** [comment]
@dergoegge Thanks for the ping.

> makes operations such as deleting them from multiple threads difficult 

Why is this a goal?

I'm quite nervous about these changes. I've spoken with @dergoegge about this a few times now, but I'm afraid my information and opinions on the code involved are quite dated at this point (2017 to be exact :), so I haven't had much to contribute. 

I'll say though, any changes or refactors for future changes to the threading model should be very well justified upfront imo. I remember even being uneasy about #10738 at the time, in fact I might have even talked myself out of rebasing it for merge.

**Review by dergoegge — CHANGES_REQUESTED** [review]

**Comment by dergoegge** [comment]
>>  makes operations such as deleting them from multiple threads difficult
>
> Why is this a goal?

I missed this in the OP and think that this should not be a goal or should be well justified.

I like getting rid of the custom ref counting for other reasons. Mainly, because it's annoying and bug prone.

Something I have been working on is hiding CNode from the public net interface and turning the connection manager into a NodeId based manager, e.g. `CConnman::PushMessage(CNode*, CSerializedNetMsg&&)` becomes  `CConnman::PushMessage(NodeId, CSerializedNetMsg&&)`. This leads to situations where the CConnman needs to hold `m_nodes_mutex` while acquiring the right CNode to call `AddRef` and then later `Release` (once finished with the CNode). While that avoids holding `m_nodes_mutex` the entire time, it is pretty messy/bug-prone while we are using the custom ref-counting and smart pointers would be a good improvement there.

**Comment by willcl-ark** [comment]
> @dergoegge Thanks for the ping.
> 
> > makes operations such as deleting them from multiple threads difficult
> 
> Why is this a goal?
> 
> I'm quite nervous about these changes. I've spoken with @dergoegge about this a few times now, but I'm afraid my information and opinions on the code involved are quite dated at this point (2017 to be exact :), so I haven't had much to contribute.
> 
> I'll say though, any changes or refactors for future changes to the threading model should be very well justified upfront imo. I remember even being uneasy about #10738 at the time, in fact I might have even talked myself out of rebasing it for merge.

Thanks, and I agree that after working on this I'm not sure this will ever extend to being  threadsafe (todo: update PR description here). For context https://github.com/bitcoin/bitcoin/pull/27912 is how I came to be looking at this.

I feel like the model as (intended to be) implemented here is easier to reason about as refcounting is taken care of for us via the smart pointers and node cleanup only happens (automatically) once we have a single final reference in `m_nodes_disconnected` and we drop it: once moved from `m_nodes` the node won't be included in any future iterations of that container and it waits to be gracefully dropped from `m_nodes_disconnected` and cleaned up when appropriate.

@dergoegge also left some comments on the current implementation of the approach while I was replying here, which I will consider and reply to seperately :)

File: src/net.h [hunk 1]
```diff
@@ -798,11 +798,11 @@ class CConnman
     // alias for thread safety annotations only, not defined
     RecursiveMutex& GetNodesMutex() const LOCK_RETURNED(m_nodes_mutex);
 
-    bool ForNode(NodeId id, std::function<bool(CNode* pnode)> func);
+    bool ForNode(NodeId id, std::function<bool(CNodeRef pnode)> func);
```
Comment (ajtowns) [inline src/net.h:787]: I don't think this makes sense in general? If you have a shared_ptr, you should normally just pass the reference to the object (`CNode&`) if you've checked it's not null, or the pointer itself (`CNode*`).

cf http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-smart

**Comment by willcl-ark** [comment]
Closing for now

**Pull request closed without merging**